drivers: pwm: header cleanup

gmarull opened the pull request:
This PR includes some prep-work before introducing `pwm_dt_spec` facilities, as some other APIs do. It mainly contains:

- Doxygen cleanup/improvements
- Code formatting
- Minor code improvements (error handling, includes...)

gmarull committed:
drivers: pwm: improve some Doxygen aspects

- remove redundant @typedef
- remove redundant @a (function with () automatically resolves)
- use @see instead of "See..."

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: hide API ops from public documentation

Hide all API ops (typedefs and struct) under INTERNAL_HIDDEN. This is
information required by drivers only, not API clients.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: make API dereferencing homogeneous

Always use const pointer (dev->api is const)

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: improve error handling code and docs

Do not convert get_cycles_per_sec() failures to -EIO, just propagate the
failure code since the failure is not handled. As a consequence, all
Doxygen docstrings have been updated.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: use essentially boolean expressions

s/if (err)/if (err < 0)

Ref. MISRA rule 10.1.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: group all capture APIs together

Group all capture APIs together for better readability. Previous to this
patch, get_cycles() as placed at the end.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: improve formatting

Improve formatting (parameters alignment, declarations, etc.).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: improve include list

- sort includes (libc, zephyr)
- make header self-contained
- remove redundant includes

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: use UINT32_MAX for overflow checks

Some values are limited to 32-bit, UINT32_MAX can be used to check if
they overflow.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

henrikbrixandersen commented:
> I think we should be consistent everywhere, that is, rename the API to be `pwm_channel_...` as well (with `pwm_pin_...` deprecation as it is stable).

I agree. Did you want to include this change + deprecation in this PR?

carlescufi reviewed (dismissed):
The current set of changes is fine with me.

gmarull committed:
drivers: pwm: improve some Doxygen docstrings

Make some docstrings a bit more clear/concise.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: add Doxygen in/out tags to parameters

Indicate wether pointer parameters are input or output. This information
is useful for API users.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> > I think we should be consistent everywhere, that is, rename the API to be `pwm_channel_...` as well (with `pwm_pin_...` deprecation as it is stable).
> 
> I agree. Did you want to include this change + deprecation in this PR?

I've dropped name changes from this PR, will open a follow-up PR with the proposal.

gmarull commented on include/drivers/pwm.h:
```diff
@@ -21,10 +21,13 @@
-#include <sys/math_extras.h>
+#include <stdint.h>
+
+#include <autoconf.h>
```
I've just tried to make the header self-contained, I'm not sure if `autoconf.h` is supposed to be part of some other "catch-all" header, do you know?

gmarull replied:
ok, figured it out, works using `-imacros`, so it doesn't have to be included. removed

anangl commented on include/drivers/pwm.h:
```diff
@@ -232,7 +231,7 @@ static inline int pwm_pin_configure_capture(const struct device *dev,
 					    void *user_data)
 {
-	const struct pwm_driver_api *api = (struct pwm_driver_api *)dev->api;
+	const struct pwm_driver_api *api = dev->api;
```
These type castings are not redundant. They are needed when this header file is included from a C++ source file.

gmarull replied:
Argh, forgot about C++ :-(

nashif commented on include/drivers/pwm.h:
```diff
@@ -21,10 +21,12 @@
+
 #include <device.h>
+#include <sys/math_extras.h>
+#include <toolchain.h>
```
what do we need from here?
